### PR TITLE
feat(HEOS): superancillary happy path for D+{H,S,U} flash

### PR DESCRIPTION
**Pull request by ibell**
## Summary

Adds a superancillary **"happy path"** to `FlashRoutines::HSU_D_flash` (density + {H,S,U} flash) for pure fluids, falling through to the legacy ancillary path on any failure (mixtures, pseudo-pure, fluids without a superancillary). Tracked in `bd CoolProp-j3n`.

This is the long-standing flaky family behind #1965, #2154, #2157, #1698, #1054, #2022, #2685, #2426, #2042 (D+{H,S,U} / energy-conservation inputs).

## Robustness — strictly better, no regressions

A/B vs master-equivalent (legacy + superancillaries on, via the new `COOLPROP_DISABLE_SUPERANC_HSU_D` kill-switch), over a dense (log-p, linear-T) sweep across **every** superancillary fluid (~5.5M states):

- **Regressions (fail happy / pass legacy): 0**
- **Fixes (fail legacy / pass happy): 2,839** across 54 fluids
- The happy path's residual failure set is a strict **subset** of legacy's.

The happy path is **self-validating** (`committed_ok` gate): it never *returns* a state that fails to reproduce the inputs; on failure it restores the caller's inputs and falls through to legacy.

## Design

- **1D-in-T two-phase residual** reads saturated ρ off the superancillary (`eval_sat`), replacing the legacy inner 2D `saturation_D_pure` solve.
- **Interval classification by dome membership** (not root count) handles the ρ≈ρ_crit collapse and the water/heavy-water liquid **density anomaly** uniformly; never hands the single-phase solver a dome-interior bracket.
- Two-phase quality clamped at the saturation boundary; spurious single-phase crossings rejected when the implied quality leaves [0,1].

## Performance

The two-phase residual reads saturated h/s/u straight off the **caloric superancillary** (~150 ns/call vs ~2.7 µs with EOS `keyed_output` — profiled). A short **full-EOS secant polish** (default on, new config `HSU_D_TWOPHASE_EOS_POLISH`) cleans the ~1e-8 superancillary deviation to EOS-exact.

Two-phase D+S / D+U (Release, polished, vs legacy): **~2.7× / ~2.5×**; near-critical-density single-phase ~3× (avoids legacy's `saturation_D_pure` ω-halving retry loop). With the polish off: ~8×.

A **native U caloric superancillary** is now built alongside H and S (`ensure_HSU_under_lock`; storage and the `'U'` build path already existed). U shares H's reference-frame shift (`u = h − p/ρ`, p/ρ offset-independent), so all three of D+{H,S,U} take the same uniform fast path — a **net simplification** (removes the U-specific gate and the `h − p/ρ` reconstruction) that brings D+U up to H/S speed. Runtime-built like H/S; **no `fastchebpure`/JSON regeneration**.

Escape hatch: `COOLPROP_DISABLE_SUPERANC_HSU_D` forces the legacy path.

## Tests

`src/Tests/CoolProp-Tests-HSU_D.cpp` (tag `[HSU_D]`): characterization suite mined from the issues above, water density-anomaly cases, polish-config toggle, plus opt-in `[HSU_D_ptsweep]` (dense all-fluid sweep, env-tunable grid) and `[HSU_D_bench]`/`[HSU_D_prof]` timing/profiling.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Improved HSU_D flash: faster, more robust density + enthalpy/entropy/internal‑energy solves for pure fluids with an optional two‑phase EOS polish (new configuration toggle, default: true). Caloric ancillary handling consolidated for better stability and accuracy.

* **Tests**
  * Added comprehensive regression and benchmark suite covering saturation boundaries, two‑phase interiors, near‑critical regions, fluid‑specific edge cases, and performance comparisons.

[Review Change Stack](https://app.coderabbit.ai/change-stack/CoolProp/CoolProp/pull/2995?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/Backends/Helmholtz/FlashRoutines.cpp [hunk 1]
```diff
@@ -1444,6 +1520,281 @@ void FlashRoutines::HSU_D_flash(HelmholtzEOSMixtureBackend& HEOS, parameters oth
...
+        // H/S/U, use the superancillary saturation-density roots to robustly
+        // decide single- vs two-phase and to bracket the temperature solve.
+        // This avoids the saturation_D_pure instability that makes the legacy
+        // routine fail near the critical density, and clamps the quality at the
+        // saturation boundary.  Any failure here falls through to the legacy
+        // ancillary "sad path" below.  See CoolProp-j3n (ported from saD_HSU).
+        //
+        // Escape hatch: COOLPROP_DISABLE_SUPERANC_HSU_D (read once) forces the
+        // legacy path -- a field kill-switch and an A/B regression-test handle.
+        static const bool superanc_hsu_d_disabled = (std::getenv("COOLPROP_DISABLE_SUPERANC_HSU_D") != nullptr);
```
Comment (ibell) [inline src/Backends/Helmholtz/FlashRoutines.cpp:1535]: If this even necessary if the disable-via-config is active?

File: src/Backends/Helmholtz/FlashRoutines.cpp [hunk 2]
```diff
@@ -1444,6 +1520,281 @@ void FlashRoutines::HSU_D_flash(HelmholtzEOSMixtureBackend& HEOS, parameters oth
...
+                            double Qd_final;
+                            if (use_ca && get_config_bool(HSU_D_TWOPHASE_EOS_POLISH)) {
+                                solver_resid_2phase resid_eos(HEOS, HEOS._rhomolar, other, value, superanc, /*use_ca=*/false, ca_key, value_cache);
+                                double t0 = Tsol, r0 = resid_eos.call(t0);
+                                // Probe just above Tsol, but keep the second secant seed
+                                // strictly inside the two-phase bracket [Ta, Tb]: when the
+                                // raw solve lands on a boundary root, Tsol*(1+1e-7) can step
+                                // past Tb and make resid_eos.call() fail, needlessly kicking
+                                // a valid happy-path state back to the legacy path.
+                                double t1 = Tsol * (1.0 + 1e-7);
```
Comment (ibell) [inline src/Backends/Helmholtz/FlashRoutines.cpp:1694]: Should this be a Newton step with analytic derivatives?

File: src/Backends/Helmholtz/FlashRoutines.cpp [hunk 3]
```diff
@@ -1444,6 +1520,281 @@ void FlashRoutines::HSU_D_flash(HelmholtzEOSMixtureBackend& HEOS, parameters oth
...
+                            } else {
+                                return false;
+                            }
+                            // EOS polish: the fast path read the saturated caloric
+                            // values off the superancillary, which carries a ~1e-8
+                            // deviation from the EOS.  T_sol is an excellent seed, so a
+                            // couple of secant steps on the EOS-caloric residual (full
+                            // EOS h/s/u of each phase) clean the solution up to EOS
+                            // precision at a few EOS evaluations -- vs the ~12 the fast
+                            // path saved.  U already uses the EOS, so nothing to polish.
```
Comment (ibell) [inline src/Backends/Helmholtz/FlashRoutines.cpp:1681]: This comment is wrong about U

**Comment by ibell** [comment]
@coderabbitai Re: the secant bracket-clamping suggestion at `FlashRoutines.cpp` (keep later iterates inside `[Ta, Tb]`) — **not applied; verified it regresses behavior.** I implemented the proposed `|| tn <= Ta || tn >= Tb` break and it broke 13 assertions in the #2486 saturation-boundary characterization cases (n-Propane, R134a), dropping previously-passing happy-path solves to the legacy path (which then throws `saturation_D_pure solver abs error`).

Why the premise doesn't hold here:

1. **The polish residual is valid far outside `[Ta, Tb]`.** `solver_resid_2phase::call(T)` reads saturated properties off `superanc.eval_sat(T, …)`, which is valid across the *entire* saturation curve (T_triple…T_crit). It does not throw outside `[Ta, Tb]`; it only throws on a non-finite residual, which is already guarded by the `std::isfinite(r1)` check / the `isfinite(tn)` break. So "the next `resid_eos.call(t1)` will throw" is not the case for this residual.

2. **`[Ta, Tb]` is the single-phase solver's shy-of-dome bracket, not a hard domain for the caloric polish.** Those bounds are deliberately set "just shy of" the dome to keep the *single-phase* solver out of the dome interior. For saturation-boundary inputs the EOS-exact saturation T sits right at / just outside that shy margin, so constraining the secant to `(Ta, Tb)` actively prevents convergence to the correct root.

3. **No UB / wrong-result risk to harden against.** The path self-validates via `committed_ok()`; a bad polish would at worst fall back to legacy, never return a wrong state.

Leaving the secant unconstrained (current code) is correct. Closing this one as won't-fix with the above rationale.
